File: src/crewlet/agent/skills/guard.py
"""Required-skill guard — load-before-use enforcement for tool skills.

A tool skill's body carries practices an agent should read *before*
touching the tools the skill's trigger covers (workflow constraints,
mention markup, delegation conventions) — and LLMs routinely skip the
``load_tool_skill`` call and go straight for the tool.  This module
closes that gap in code, not in prompts.  Skills are enforced by
default (``required: true``); operators mark orientation / hint-grade
content ``required: false`` to keep it advisory.

One :class:`SkillGuard` is attached to the phase's ``ToolSurface``
(``surface.skill_guard``) by each phase runner that carries domain
tools — Plan, Execute, and Sub-agent.  The shared dispatch gate
(:func:`crewlet.agent.llm_loop.execute_tool`) consults it on every
tool call:

- a call to a tool covered by a required-but-unloaded skill is
  rejected with an instructive error naming the exact
  ``load_tool_skill(key=...)`` call to make, and a
  ``phase.tool_skill_blocked`` event is published for operators;
- a successful ``load_tool_skill`` call records the key, unlocking
  the covered tools for the rest of the session.

**Session scope.**  "Loaded" is tracked per LLM session — one phase's
message history — not per turn.  Plan, Execute, and each sub-agent run
on separate message histories, so a body loaded during Plan is not in
the Execute LLM's context; each session must load the skill itself.
Round-cap extension loops continue the same message history *and* the
same surface object, so the loaded set carries across extensions.
``self_iterate`` starts fresh phase sessions and therefore fresh
guards.

The guard never fires for sessions that cannot recover: it refuses to
arm when ``load_tool_skill`` is absent from the surface
(:func:`build_skill_guard` returns ``None``), and engine plumbing
tools (the unlock itself, discovery meta-tools, phase-contract
submitters) are exempt so a misauthored trigger can't brick a phase.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from crewlet._logging import get_logger
from crewlet.agent.skills.models import Phase, PromptSkill, TriggerContext
from crewlet.agent.skills.registry import PromptSkillRegistry
from crewlet.events.types import ToolSkillGuardBlocked
from crewlet.tools.protocol import ToolResult

logger = get_logger("agent.skills.guard")

GUARD_EXEMPT_TOOLS: frozenset[str] = frozenset(
    {
        # The unlock itself — gating it would deadlock the session.
        "load_tool_skill",
        # Discovery meta-tools: the LLM may need them to even reach the
        # guarded tool (list server tools, activate). Blocking them
        # adds rounds without protecting anything.
        "activate_tool",
        "list_mcp_server_tools",
        # Phase-contract submitters: blocking these bricks the phase
        # (the rescue paths force them with tool_choice="required").
        "submit_plan",
        "submit_review",
    }
)
"""Tools the guard never blocks, regardless of skill triggers."""


@dataclass
class SkillGuard:
    """Per-session load-before-use enforcement state.

    Built once per phase session by :func:`build_skill_guard` and
    attached to the phase's ``ToolSurface``.  ``trigger_ctx`` is the
    same :class:`TriggerContext` the phase's prompt catalogue was
    built from, so the guard gates exactly the skills the LLM was
    shown (catalogue and enforcement can't disagree).

    Event metadata (``event_queue`` / ``agent_id`` / ... ) is optional
    — guards built in tests without a queue simply skip publishing.
    """

    registry: PromptSkillRegistry
    phase: Phase
    trigger_ctx: TriggerContext
    event_queue: Any = None
    agent_id: str = ""
    agent_role: str = ""
    turn_id: str = ""
    iteration: int = 0
    loaded_keys: set[str] = field(default_factory=set)

    def missing_for(self, tool: Any) -> list[PromptSkill]:
        """Required skills covering ``tool`` not yet loaded this session."""
        name = getattr(tool, "name", "")
        if not name or name in GUARD_EXEMPT_TOOLS:
            return []
        skills = self.registry.required_skills_for_tool(
            phase=self.phase,
            tool_name=name,
            server_name=getattr(tool, "server_name", "") or "",
            ctx=self.trigger_ctx,
        )
        return [s for s in skills if s.key not in self.loaded_keys]

    async def check_tool(self, tool: Any) -> ToolResult | None:
        """Gate one tool call.

        Returns ``None`` when the call may proceed, or a failed
        :class:`ToolResult` telling the LLM which skill(s) to load
        first.  Publishes a ``phase.tool_skill_blocked`` event on every
        block so operators can see agents skipping required practices.
        """
        missing = self.missing_for(tool)
        if not missing:
            return None
        name = getattr(tool, "name", "")
        keys = [s.key for s in missing]
        logger.info(
            "tool_skill_guard_blocked",
            tool_name=name,
            skill_keys=keys,
            phase=str(self.phase),
            agent_id=self.agent_id,
            turn_id=self.turn_id,
        )
        await self._publish_blocked(tool_name=name, skill_keys=keys)
        return ToolResult(
            success=False, error=_block_message(name, missing, self.registry)
        )

    def observe(self, name: str, arguments: dict[str, Any], success: bool) -> None:
        """Record a completed tool call.

        Only successful ``load_tool_skill`` calls matter: the loaded
        key unlocks every tool that skill covers for the rest of this
        session.
        """
        if name != "load_tool_skill" or not success:
            return
        key = str(arguments.get("key", "")).strip()
        if not key:
            return
        self.loaded_keys.add(key)
        logger.debug(
            "tool_skill_guard_recorded_load",
            key=key,
            phase=str(self.phase),
            agent_id=self.agent_id,
        )

    async def _publish_blocked(self, *, tool_name: str, skill_keys: list[str]) -> None:
        if self.event_queue is None:
            return
        event = ToolSkillGuardBlocked(
            source=self.agent_role,
            agent_id=self.agent_id,
            role=self.agent_role,
            phase=str(self.phase),
            tool_name=tool_name,
            skill_keys=list(skill_keys),
            turn_id=self.turn_id,
            iteration=self.iteration,
        )
        try:
            await self.event_queue.publish(f"crewlet.events.{event.type}", event)
        except Exception:
            logger.exception(
                "tool_skill_blocked_publish_failed",
                tool_name=tool_name,
                skill_keys=skill_keys,
            )


def _block_message(
    tool_name: str,
    missing: list[PromptSkill],
    registry: PromptSkillRegistry,
) -> str:
    """Instructive rejection the LLM can act on in the next round.

    Summaries are rendered through the registry so any ``${var}``
    references resolve here too — this message goes straight to the LLM.
    """
    loads = "; then ".join(f"`load_tool_skill(key='{s.key}')`" for s in missing)
    listing = "\n".join(
        f"- '{s.key}': {' '.join(registry.render(s.summary).split())}" for s in missing
    )
    plural = "skills" if len(missing) > 1 else "skill"
    return (
        f"Tool '{tool_name}' is gated behind required tool {plural} you "
        f"have not loaded in this session:\n{listing}\n\n"
        f"Call {loads} to read the required practice(s), then retry "
        f"'{tool_name}'. Loading is needed once per session; after that "
        "the tool works normally."
    )


def build_skill_guard(
    *,
    registry: Any,
    phase: Phase,
    surface: Any,
    mcp_servers: set[str],
    event_queue: Any = None,
    agent_id: str = "",
    agent_role: str = "",
    turn_id: str = "",
    iteration: int = 0,
) -> SkillGuard | None:
    """Build the guard for one phase session, or ``None`` when it
    cannot or should not arm.

    - ``registry is None`` — tool skills not configured for this
      engine; nothing to enforce.
    - ``load_tool_skill`` not in ``surface`` — the session has no way
      to satisfy the guard; arming it would soft-lock the LLM in a
      block-retry loop. (Plan carries it as a meta-tool, Execute as an
      always-on, Sub-agent surfaces always include it — so this is a
      defensive rail for non-standard surfaces, not a normal path.)

    The trigger context mirrors what the phase's prompt catalogue was
    built from: the surface's full catalogue (activation universe) for
    Plan / Execute, falling back to the active tool list for Sub-agent
    surfaces (which carry no catalogue), plus the role's MCP servers.
    """
    if registry is None:
        return None
    if not surface.has("load_tool_skill"):
        logger.warning(
            "skill_guard_disabled_no_loader",
            phase=str(phase),
            agent_id=agent_id,
            turn_id=turn_id,
        )
        return None
    tools = set(surface.catalogue_names()) or set(surface.names)
    return SkillGuard(
        registry=registry,
        phase=phase,
        trigger_ctx=TriggerContext(
            tools=frozenset(tools),
            mcp_servers=frozenset(mcp_servers),
        ),
        event_queue=event_queue,
        agent_id=agent_id,
        agent_role=agent_role,
        turn_id=turn_id,
        iteration=iteration,
    )


def skill_guard_for_turn(
    *,
    registry: Any,
    phase: Phase,
    surface: Any,
    turn: Any,
    event_queue: Any = None,
) -> SkillGuard | None:
    """:func:`build_skill_guard` with the per-turn fields derived from a
    :class:`~crewlet.agent.turn_context.TurnContext`.

    The three phase runners (Plan / Execute / Sub-agent) all derive the
    same five arguments from their turn context — the role's MCP-server
    set and the agent / turn identity for the blocked event.  Deriving
    them here keeps the three call sites to one expression and removes
    the risk of the derivations drifting apart.  ``turn`` is typed
    ``Any`` to avoid coupling this module to the turn-context import
    graph; it duck-types ``TurnContext``.
    """
    return build_skill_guard(
        registry=registry,
        phase=phase,
        surface=surface,
        mcp_servers=set(getattr(turn.agent.definition.role, "mcp_env", {}) or {}),
        event_queue=event_queue,
        agent_id=turn.agent.id_str,
        agent_role=turn.agent.role_name,
        turn_id=turn.turn_id,
        iteration=turn.iteration,
    )


__all__ = [
    "GUARD_EXEMPT_TOOLS",
    "SkillGuard",
    "build_skill_guard",
    "skill_guard_for_turn",
]

File: internal/agent/skills/guard.go
package skills

import (
	"fmt"
	"maps"
	"slices"
	"strings"
	"sync"

	"github.com/crewlet/crewlet/internal/agent/prompts"
)

// Load-before-use, enforced in code rather than in prose.
//
// A skill's body carries what an agent should read BEFORE touching the tools
// its trigger covers — a workflow constraint, the mention markup, a
// delegation convention — and models routinely skip the load and go straight
// for the tool. Asking harder in the prompt does not fix that; refusing the
// call does.
//
// # Session scope, not turn scope
//
// "Loaded" is tracked per LLM SESSION — one phase's message history — because
// that is what determines whether the body is actually in the model's
// context. Plan, Execute and each sub-agent run on separate histories, so a
// body loaded during Plan is genuinely not in front of the executor. A
// round-cap extension continues the same history and the same surface, so
// the loaded set carries across it; a self_iterate starts a fresh session
// and therefore a fresh guard, which is correct rather than annoying.

// ExemptTools are never blocked, whatever a trigger says.
//
// Three kinds, each for a reason that is about DEADLOCK rather than policy:
// the unlock itself (gating it would make the session unrecoverable), the
// discovery meta-tools (the model may need them to reach the guarded tool at
// all, so blocking them adds rounds and protects nothing), and the phase
// submitters (blocking one bricks the phase — the rescue paths force them).
//
// A misauthored trigger can then cost a phase some tools; it can never cost
// the phase.
var ExemptTools = []string{
	LoaderTool,
	"activate_tool",
	"list_mcp_server_tools",
	"submit_work",
	"submit_review",
	"mark_onboarded",
}

// Blocked is a refused call, and what the model must do to proceed.
type Blocked struct {
	Tool string

	// Skill is the key to load, and Title is what it is about — both, so
	// the message names the exact call to make AND says why it is worth
	// making. A bare key reads as an obstacle.
	Skill string
	Title string
}

// Error renders the refusal the model sees.
//
// It NAMES THE EXACT CALL, because the error is the recovery path and a
// model told only "you must load a skill" spends a round discovering which.
// The discovery path is the catalogue, which already carries the marker; by
// the time this fires the model has skipped it.
func (b Blocked) Error() string {
	about := ""
	if b.Title != "" {
		about = " (" + b.Title + ")"
	}
	return fmt.Sprintf("%s is covered by the required tool skill %q%s, which "+
		"has not been loaded in this session. Call load_tool_skill(key=%q) "+
		"first, then retry %s.", b.Tool, b.Skill, about, b.Skill, b.Tool)
}

// Guard is one session's load-before-use state.
type Guard struct {
	registry *Registry
	phase    prompts.Phase
	surface  prompts.Surface

	mu     sync.Mutex
	loaded map[string]bool
}

// NewGuard arms the guard for one phase session, or returns nil.
//
// NIL WHEN THE SESSION CANNOT RECOVER: with no loader on the surface there
// is no way to satisfy the guard, so arming it would refuse tools the model
// has no path to unlock. Nil is also the answer for a phase with no required
// skills matching, which is the overwhelmingly common case and costs nothing
// on every call.
//
// The surface it is given is the SAME one the catalogue was built from, so
// what is enforced and what the model was shown cannot disagree.
func NewGuard(r *Registry, phase prompts.Phase, surface prompts.Surface) *Guard {
	if r == nil || !slices.Contains(surface.Tools, LoaderTool) {
		return nil
	}
	// REVIEW IS EXEMPT ENTIRELY: it has no domain tools and no loader, so
	// there is nothing to gate and a marker there would point at a tool
	// the reviewer does not have. The prompt package makes the same
	// exception for the catalogue's marker, from the same reasoning.
	if phase == prompts.PhaseReview {
		return nil
	}
	required := false
	for _, s := range r.Matching(phase, surface) {
		if s.Required {
			required = true
			break
		}
	}
	if !required {
		return nil
	}
	return &Guard{registry: r, phase: phase, surface: surface, loaded: map[string]bool{}}
}

// Restore re-arms a guard with the keys a previous slice of the SAME session
// already loaded.
//
// The one legitimate way to hand a guard somebody else's loads, and it exists
// for exactly one caller: an executor that suspended on a detached coding run
// is re-entered as the same message history days later, so the bodies it
// loaded before the suspend are still in front of the model. A guard rebuilt
// empty would block the very tools that session already unlocked, and the
// model would be told to load a skill it can see in its own transcript.
func (g *Guard) Restore(keys []string) {
	if g == nil {
		return
	}
	g.mu.Lock()
	defer g.mu.Unlock()
	for _, key := range keys {
		if key != "" {
			g.loaded[key] = true
		}
	}
}

// LoadedKeys is what this session has loaded, sorted.
//
// Sorted because it is persisted into a suspended conversation's state, and a
// blob whose key order came from map iteration would differ on every write of
// the same fact.
func (g *Guard) LoadedKeys() []string {
	if g == nil {
		return nil
	}
	g.mu.Lock()
	defer g.mu.Unlock()
	return slices.Sorted(maps.Keys(g.loaded))
}

// Loaded records a successful load, unlocking the tools that skill covers.
func (g *Guard) Loaded(key string) {
	if g == nil || key == "" {
		return
	}
	g.mu.Lock()
	defer g.mu.Unlock()
	g.loaded[key] = true
}

// LoaderTool is the tool whose successful call unlocks a skill.
const LoaderTool = "load_tool_skill"

// Observe implements [tools.Guard]: it watches for the unlock.
//
// Reading the key back off the ARGUMENTS rather than the result, because the
// result is prose for the model and parsing a key out of it would be
// guessing at a format nothing promises. The arguments are what the model
// asked for and what the tool resolved.
func (g *Guard) Observe(tool string, args map[string]any) {
	if g == nil || tool != LoaderTool {
		return
	}
	key, _ := args["key"].(string)
	g.Loaded(strings.TrimSpace(key))
}

// Check implements [tools.Guard]: it reports why a call must be refused, or
// "" to let it through.
//
// The SERVER is passed alongside the tool because a trigger's server leaf
// covers every tool that server publishes, and a tool name alone cannot say
// which server published it.
//
// A STRING rather than an error value, because the tool layer hands the
// answer straight to the model and an error's text is not something the tool
// layer should have to know how to render.
func (g *Guard) Check(tool, server string) string {
	if blocked := g.Blocking(tool, server); blocked != nil {
		return blocked.Error()
	}
	return ""
}

// Blocking is Check with the structure kept, for the caller that publishes
// telemetry about what was refused and needs the skill key rather than the
// sentence.
func (g *Guard) Blocking(tool, server string) *Blocked {
	if g == nil || slices.Contains(ExemptTools, tool) {
		return nil
	}
	loaded := g.loadedSet()

	for _, s := range g.registry.Matching(g.phase, g.surface) {
		if !s.Required || loaded[s.Key] || !s.Trigger.Covers(tool, server) {
			continue
		}
		return &Blocked{Tool: tool, Skill: s.Key, Title: s.Title}
	}
	return nil
}

// Pending are the required skills this session has not yet loaded, sorted.
//
// For telemetry and for a phase record: an operator looking at a turn that
// stalled on a guard needs to see which skill it was waiting for, and the
// blocked call alone names only the first one.
func (g *Guard) Pending() []string {
	if g == nil {
		return nil
	}
	loaded := g.loadedSet()

	var out []string
	for _, s := range g.registry.Matching(g.phase, g.surface) {
		if s.Required && !loaded[s.Key] {
			out = append(out, s.Key)
		}
	}
	slices.Sort(out)
	return out
}

// Render substitutes a skill's variables for a body about to be loaded.
func (g *Guard) Render(text string) string {
	if g == nil {
		return text
	}
	return g.registry.Render(text)
}

// loadedSet copies the loaded keys out from under the lock.
//
// COPIED rather than held, so the registry walks in Check and Pending run
// outside this lock: those walks take the REGISTRY's lock, and a guard
// holding its own while waiting on another is the shape a deadlock is made
// of.
func (g *Guard) loadedSet() map[string]bool {
	g.mu.Lock()
	defer g.mu.Unlock()
	return maps.Clone(g.loaded)
}
